File: 62-unique-paths/62-unique-paths.py
class Solution:
    def solve(self,m,n,memo):
        if m==1 or n ==1:
            return 1
        key=f"{m}-{n}"
        if key in memo:
            return memo[key]
        memo[key]=self.solve(m-1,n,memo)+self.solve(m,n-1,memo)
        return memo[key]
        
    def uniquePaths(self, m: int, n: int) -> int:
        
        return self.solve(m,n,{})
        

File: 62-unique-paths/62-unique-paths.java
class Solution {
    
    private int recSolve(int m,int n,HashMap<String,Integer> memo)
    {
        
        
        if(m==0 || n==0)
            return 0;
        else if (m==1 || n==1)
            return 1;
        
        String key=m+"-"+n;
        
        
        if(memo.containsKey(key))
            return memo.get(key);
        int downPath=recSolve(m-1,n,memo);
        int rightPath=recSolve(m,n-1,memo);
        
        memo.put(key,downPath+rightPath);
        return memo.get(key);
        
        
    }
    public int uniquePaths(int m, int n) {
        
        HashMap<String,Integer> memo=new HashMap<>();
        
        
        return recSolve(m,n,memo);
        
    }
}